pkg/metrics/exporter.go

```diff
@@ -0,0 +1,81 @@
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/version"
+)
+
+const (
+	exporterName     = "kubernikus_exporter"
+	metricsNamespace = "kubernikus"
+	defaultURI       = "0.0.0.0"
+	metricsPort      = 9091
+	metricsPath      = "metrics"
+)
+
+type Exporter struct {
+	URI string
+
+	klusterMetrics, nodePoolMetrics map[string]*prometheus.Desc
+}
+
+func NewExporter(uri string) *Exporter {
+	if uri == "" {
+		uri = defaultURI
+	}
+	return &Exporter{
+		URI: uri,
+		klusterMetrics: map[string]*prometheus.Desc{
+			"klusterTotal": newKlusterMetric("kluster_instances_total", "total number of klusters", []string{"kluster_phase", "domain_id", "project_id"}),
+		},
+		nodePoolMetrics: map[string]*prometheus.Desc{
+			"nodePoolTotal": newNodePoolMetric("node_pools_total", "total number of nodepools", []string{"status", "kluster_id"}),
+			"nodePoolSize":  newNodePoolMetric("node_pool_size_total", "total size of nodepools", []string{"status", "kluster_id"}),
+		},
+	}
+}
+
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range e.klusterMetrics {
+		ch <- metric
+	}
+	for _, metric := range e.nodePoolMetrics {
+		ch <- metric
+	}
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+
+  //TODO list klusters via API and derive the following metrics
```

Comment:
Do we want to go as described above or directly integrate in controllers?

Reply:
We can put it in the controllers as whitebox metrics. Let's do it there instead of carrying around another binary that needs transport, supervision and stuff. And isn't even able to look into the applications.

I'd say we make it easy and cheap to put instrumentation all over the place. Complement and reduce the required logging...